packages/tpetra/core/src/Tpetra_Details_lclMultiply.hpp

```diff
@@ -0,0 +1,149 @@
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_DETAILS_LCLMULTIPLY_HPP
+#define TPETRA_DETAILS_LCLMULTIPLY_HPP
+
+/// \file Tpetra_Details_lclMultiply.hpp
+/// \brief Declaration and definition of Tpetra::Details::lclMultiply, an
+///   implementation detail of Tpetra::MultiVector.
+///
+/// \warning This file, and its contents, are implementation details
+///   of Tpetra, and may change or disappear at any time.
+
+#include "Kokkos_ArithTraits.hpp"
+#include "KokkosBlas.hpp"
+#include "Tpetra_Details_Profiling.hpp"
+#include "Kokkos_InnerProductSpaceTraits.hpp"
+
+namespace Tpetra {
+namespace Details {
+
+template<class XMV,
+         typename std::enable_if<std::is_same<typename XMV::value_type,typename Kokkos::Details::InnerProductSpaceTraits<typename XMV::value_type>::dot_type>::value >::type* = nullptr >
+void
+lclMultiply(const Teuchos::ETransp transA,
+            const Teuchos::ETransp transB,
+            typename XMV::value_type alpha,
+            const XMV& A_lcl,
+            const XMV& B_lcl,
+            typename XMV::value_type beta,
+            XMV& C_lcl)
+{
+  using Teuchos::CONJ_TRANS;
+  using Teuchos::NO_TRANS;
+  using Teuchos::TRANS;
+  using Kokkos::ALL;
+  using Kokkos::subview;
+  using VT = typename XMV::value_type;
+  using ATS = Kokkos::ArithTraits<VT>;
+
+
+  const char ctransA = (transA == NO_TRANS ? 'N' :
+                        (transA == TRANS ? 'T' : 'C'));
+  const char ctransB = (transB == NO_TRANS ? 'N' :
+                        (transB == TRANS ? 'T' : 'C'));
+
+  const size_t zero = 0;
+  const VT zeroVT = ATS::zero();
+  const VT oneVT = ATS::one();
+
+
+  // The first two cases are specializations for inner products between a vector and a multivector
+  if (beta == zeroVT && transA != NO_TRANS && transB == NO_TRANS && A_lcl.extent(1) == 1) {
+    auto C_sub = subview (C_lcl, zero, ALL ());
+    ::Tpetra::Details::ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-dot");
+    KokkosBlas::dot(C_sub, A_lcl, B_lcl);
+    if (alpha != oneVT)
+      KokkosBlas::scal (C_sub, alpha, C_sub);
+  }
+  else if (beta == zeroVT && transA != NO_TRANS && transB == NO_TRANS && B_lcl.extent(1) == 1) {
+    auto C_sub = subview (C_lcl, ALL (), zero );
+    ::Tpetra::Details::ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-dot");
+    KokkosBlas::dot(C_sub, A_lcl, B_lcl);
+    if (alpha != oneVT)
+      KokkosBlas::scal (C_sub, alpha, C_sub);
+  }
+  else {
+    auto C_sub = subview (C_lcl,
+                          std::make_pair (zero, C_lcl.extent(0)),
+                          std::make_pair (zero, C_lcl.extent(1)));
+    ::Tpetra::Details::ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-gemm");
+    KokkosBlas::gemm(&ctransA, &ctransB, alpha, A_lcl, B_lcl, beta, C_sub);
```

Comment:
Do you see what I mean?  Both the "dot type same as value type" and "dot type not same as value type" branches invoke `KokkosBlas::gemm` with arguments of the same types.

Reply:
I think we figured this out -- the issue was that one branch wanted to use `dot` to do a GEMV, but that only works if `dot_type == value_type`.  I think we should just use GEMV in that case (see my comments above).